vaultlocker/shell.py

```diff
@@ -72,26 +78,39 @@ def _encrypt_block_device(args, client, config):
     key = dmcrypt.generate_key()
     block_uuid = str(uuid.uuid4()) if not args.uuid else args.uuid
     vault_path = _get_vault_path(block_uuid, config)
-
     # NOTE: store and validate key before trying to encrypt disk
     try:
-        client.write(vault_path,
-                     dmcrypt_key=key)
+        if config.get('vault', 'kv_version') == '2':
+            client.secrets.kv.v2.create_or_update_secret(vault_path,
+                                                         secret=dict(dmcrypt_key=key),
+                                                         mount_point=config.get('vault', 'backend'))
+        elif config.get('vault', 'kv_version') == '1':
+            client.secrets.kv.v1.create_or_update_secret(vault_path,
+                                                         secret=dict(dmcrypt_key=key),
+                                                         mount_point=config.get('vault', 'backend'))
     except hvac.exceptions.VaultError as write_error:
         logger.error(
             'Vault write to path {}. Failed with error: {}'.format(
                 vault_path, write_error))
         raise exceptions.VaultWriteError(vault_path, write_error)
 
     try:
-        stored_data = client.read(vault_path)
+        if config.get('vault', 'kv_version') == '2':
+            stored_data = client.secrets.kv.v2.read_secret_version(vault_path,
+                                                                   mount_point=config.get('vault', 'backend'))
+        elif config.get('vault', 'kv_version') == '1':
+            stored_data = client.secrets.kv.v1.read_secret(vault_path,
+                                                           mount_point=config.get('vault', 'backend'))
     except hvac.exceptions.VaultError as read_error:
         logger.error('Vault access to path {}'
                      'failed with error: {}'.format(vault_path, read_error))
         raise exceptions.VaultReadError(vault_path, read_error)
 
-    if not key == stored_data['data']['dmcrypt_key']:
-        raise exceptions.VaultKeyMismatch(vault_path)
+        if not key == stored_data['data']['data']['dmcrypt_key'] and config.get('vault', 'kv_version') == '2':
```

Comment:
What's with the double keying using 'data' - looks odd?

Reply:
When using a KV v2 in Vault, the Data payload is returned with two maps, one is called "data" and the other is called "metadata", because in the V2 version of a KV you can store secrets and metadata in a key, it's indeed a bit weird, but that's how it is, you'd find the same kind of data lookup in a Go implementation

Reply:
nm - read the docs